lang/wasm.cc

```diff
@@ -105,18 +107,39 @@ struct init_arg_visitor {
         }
         auto memory = std::get<wasmtime::Memory>(*memory_export);
         uint8_t* data = memory.data(store).data();
-        size_t mem_size = memory.size(store);
+        size_t mem_size = memory.size(store) * WASM_PAGE_SIZE;
         if (!param) {
             on_internal_error(wasm_logger, "init_arg_visitor does not accept null values");
         }
         int32_t serialized_size = param->size();
         if (serialized_size > std::numeric_limits<int32_t>::max()) {
             throw wasm::exception(format("Serialized parameter is too large: {} > {}", serialized_size, std::numeric_limits<int32_t>::max()));
         }
-        auto grown = memory.grow(store, sizeof(int32_t) + serialized_size); // for fitting serialized size + the buffer itself
-        if (!grown) {
-            throw wasm::exception(format("Failed to grow wasm memory to {}: {}", serialized_size, grown.err().message()));
+        auto malloc_export = instance.get(store, "_scylla_malloc");
+        if (!malloc_export) {
+            auto grown = memory.grow(store, 1 + (sizeof(int32_t) + serialized_size - 1) / WASM_PAGE_SIZE); // for fitting serialized size + the buffer itself
+            if (!grown) {
+                throw wasm::exception(format("Failed to grow wasm memory to {}: {}", serialized_size, grown.err().message()));
+            }
+            mem_size = std::move(grown).unwrap() * WASM_PAGE_SIZE;
+        } else {
+            std::vector<wasmtime::Val> malloc_argv;
+            malloc_argv.push_back(int32_t(sizeof(int32_t) + serialized_size));
+            wasmtime::Func* malloc_func = std::get_if<wasmtime::Func>(&*malloc_export);
+            if (!malloc_func) {
+                throw wasm::exception(format("Exported object {} is not a function", "malloc"));
+            }
+            auto result = malloc_func->call(store, malloc_argv);
+            if (!result) {
+                throw wasm::exception("Calling wasm function failed: " + result.err().message());
+            }
+            std::vector<wasmtime::Val> result_vec = std::move(result).unwrap();
+            if (result_vec.size() != 1) {
+                throw wasm::exception(format("Unexpected number of returned values: {} (expected: 1)", result_vec.size()));
+            }
+            mem_size = result_vec[0].i32();
```

Comment:
have you considered extracting it out to a helper function? Looks deceptively similar to the snippet below

Reply:
Done